### ISSUE TITLE
Confusion about the easy data filter

### ISSUE THREAD

**Issue by faresobeid** [XiaomiMiMo/MiMo#12]
Thanks for this great open source work!

When reading through, it seems like the initial RL data curation filters out easy samples (perfect pass rates for code and 90%+ for math). However in 3.3.2 it is mentioned that removing questions with perfect pass rates throughout training causes instability. What would be a reason for this instability and why 90%+ pass rate for math but 100% for code?

(For the instability my guess would be that it ensures the pass rate stays high/perfect when it could fluctuate if removed)

Thanks

**Comment by faresobeid** [comment]
Oh and also, how does this straightforward approach work? Does this only apply for multi epoch settings because i'm not sure how it improves efficiency if you rollout the samples for it then remove from the training data. Also for the easy data pool, again is this a multi epoch scenario as throughout rl training, easy samples are placed in the pool?

Thanks again

<img width="707" alt="Image" src="https://github.com/user-attachments/assets/b782f7ac-c99f-4c80-ba5a-c962593fa193" />

**Comment by zhao1iang-mimo** [comment]
> Oh and also, how does this straightforward approach work? Does this only apply for multi epoch settings because i'm not sure how it improves efficiency if you rollout the samples for it then remove from the training data. Also for the easy data pool, again is this a multi epoch scenario as throughout rl training, easy samples are placed in the pool?
> 
> Thanks again
> 
> <img alt="Image" width="707" src="https://private-user-images.githubusercontent.com/111092724/439122782-b782f7ac-c99f-4c80-ba5a-c962593fa193.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYwMDk0NDcsIm5iZiI6MTc0NjAwOTE0NywicGF0aCI6Ii8xMTEwOTI3MjQvNDM5MTIyNzgyLWI3ODJmN2FjLWM5OWYtNGM4MC1iYTVhLWM5NjI1OTNmYTE5My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQzMFQxMDMyMjdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YmQ1MjYzZmMxODk1YmVjNDMyZWVkNjc2Y2NhNmIzZjNhMmUyYWNmNTAxYzA0MDM0YWRlOTY2ZDNmOGQzOTZmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.4LIp5byYURgGm99jA5eGESuiImZ7T968QTd9Srv54Cw">

Thank you for your interest in our work. The easy data resampling technique is particularly useful when training for N epochs with dynamic sampling. In later training stages, over 50% of problems typically achieve fully correct solutions. Since we discard both fully correct and fully incorrect problems during training, maintaining batch size requires significant additional sampling (rollout accounts for about 80% of the training step overhead, which is very substantial).

The easy data sampling approach works by:

Tracking rollout performance to identify perfectly solved problems
Excluding these from subsequent sampling
However, we observed that some previously mastered problems may be forgotten later in training. The easy data resampling method addresses this by:

Selectively sampling from the historical perfectly-solved list
Reintroducing these problems to training
This effectively prevents catastrophic forgetting of previously learned solutions.

> Oh and also, how does this straightforward approach work? Does this only apply for multi epoch settings because i'm not sure how it improves efficiency if you rollout the samples for it then remove from the training data. Also for the easy data pool, again is this a multi epoch scenario as throughout rl training, easy samples are placed in the pool?
> 
> Thanks again
> 
> <img alt="Image" width="707" src="https://private-user-images.githubusercontent.com/111092724/439122782-b782f7ac-c99f-4c80-ba5a-c962593fa193.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYwMDk0ODAsIm5iZiI6MTc0NjAwOTE4MCwicGF0aCI6Ii8xMTEwOTI3MjQvNDM5MTIyNzgyLWI3ODJmN2FjLWM5OWYtNGM4MC1iYTVhLWM5NjI1OTNmYTE5My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQzMFQxMDMzMDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hZmRmN2YwN2NhNzZmOWQ3MzVkNzM4ZmViMzMyZDg0ZDQzNWQ2M2YwNmNjZjNiNzdiZDk2NGVhZjNiODkwZDI3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.4LivSHepbfyTfMHwD9tvyxl_yKzbgi40JLAdGVsL8Mc">

Thank you for your interest in our work. The easy data resampling technique is particularly useful when training for N epochs with dynamic sampling. In later training stages, over 50% of problems typically achieve fully correct solutions. Since we discard both fully correct and fully incorrect problems during training, maintaining batch size requires significant additional sampling (rollout accounts for about 80% of the training step overhead, which is very substantial).

The easy data sampling approach works by:

- Tracking rollout performance to identify perfectly solved problems
- Excluding these solved problems from sampling to prevent redundant rollouts

However, we observed that some previously perfectly solved problems may be forgotten later in training. The easy data resampling method addresses this by:

- Selectively sampling from the historical perfectly-solved list
- Reintroducing these problems to training
This effectively prevents catastrophic forgetting of previously learned solutions.

**Comment by FerryHuang** [comment]
> However, we observed that some previously perfectly solved problems may be forgotten later in training. The easy data resampling method addresses this by:
> 
> Selectively sampling from the historical perfectly-solved list
> Reintroducing these problems to training
> This effectively prevents catastrophic forgetting of previously learned solutions.

From my view, the dynamic sampling mechanism itself handles the so-called _catastrophic forgetting of previously learned solutions_ by **dynamically** measuring the solve-rate of a question with exactly the current policy. 

Oh I kinda understand it here. So the algo. actually drops the questions directly from the dataset once they are solved perfectly and then are re-sampled into the training ensuring they are not forgotten?

Correct me if I still got it wrong, thanks :)

**Comment by zhao1iang-mimo** [comment]
> > Oh and also, how does this straightforward approach work? Does this only apply for multi epoch settings because i'm not sure how it improves efficiency if you rollout the samples for it then remove from the training data. Also for the easy data pool, again is this a multi epoch scenario as throughout rl training, easy samples are placed in the pool?
> > Thanks again
> > <img alt="Image" width="707" src="https://private-user-images.githubusercontent.com/111092724/439122782-b782f7ac-c99f-4c80-ba5a-c962593fa193.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYwMDI0NDUsIm5iZiI6MTc0NjAwMjE0NSwicGF0aCI6Ii8xMTEwOTI3MjQvNDM5MTIyNzgyLWI3ODJmN2FjLWM5OWYtNGM4MC1iYTVhLWM5NjI1OTNmYTE5My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQzMFQwODM1NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03NWVhMTIxYTllOTE4MTcwMjM4YmY1YjU2YmJiZTRiZDdhODM1MDlkYmVjY2FkMmE4ZjZlZDM3ZjE4NzllNmI3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bXDMHQmLKxZIWQycmIcBcn0zmBQORyVViuLQ9HL2uIE">
> 
> Also confused here, the perfect-pass-questions(i.e. all samples receive 1 acc reward) should contribute no gradient to GRPO training as mentioned in DAPO's dynamic sampling. Then why they are removed first and re-sampled as training data?

the perfect-pass-questions

> From my view, the dynamic sampling mechanism itself handles the so-called _catastrophic forgetting of previously learned solutions_ by **dynamically** measuring the solve-rate of a question with exactly the current policy.
> 
> Oh I kinda understand it here. So the algo. actually drops the questions directly from the dataset once they are solved perfectly and then are re-sampled into the training ensuring they are not forgotten?
> 
> Correct me if I still got it wrong, thanks :)

You got it exactly right! That's precisely how it works - we drop the perfectly solved questions from the main dataset but keep them in a separate pool for occasional re-sampling. This way they don't dominate the training but still get revisited often enough to prevent forgetting.

**Comment by zhao1iang-mimo** [comment]
> > > Oh and also, how does this straightforward approach work? Does this only apply for multi epoch settings because i'm not sure how it improves efficiency if you rollout the samples for it then remove from the training data. Also for the easy data pool, again is this a multi epoch scenario as throughout rl training, easy samples are placed in the pool?
> > > Thanks again
> > > <img alt="Image" width="707" src="https://private-user-images.githubusercontent.com/111092724/439122782-b782f7ac-c99f-4c80-ba5a-c962593fa193.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYwMDk0NDcsIm5iZiI6MTc0NjAwOTE0NywicGF0aCI6Ii8xMTEwOTI3MjQvNDM5MTIyNzgyLWI3ODJmN2FjLWM5OWYtNGM4MC1iYTVhLWM5NjI1OTNmYTE5My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQzMFQxMDMyMjdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YmQ1MjYzZmMxODk1YmVjNDMyZWVkNjc2Y2NhNmIzZjNhMmUyYWNmNTAxYzA0MDM0YWRlOTY2ZDNmOGQzOTZmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.4LIp5byYURgGm99jA5eGESuiImZ7T968QTd9Srv54Cw">
> > 
> > 
> > Thank you for your interest in our work. The easy data resampling technique is particularly useful when training for N epochs with dynamic sampling. In later training stages, over 50% of problems typically achieve fully correct solutions. Since we discard both fully correct and fully incorrect problems during training, maintaining batch size requires significant additional sampling (rollout accounts for about 80% of the training step overhead, which is very substantial).
> > The easy data sampling approach works by:
> > 
> > * Tracking rollout performance to identify perfectly solved problems
> > * Excluding these solved problems from sampling to prevent redundant rollouts
> > 
> > However, we observed that some previously perfectly solved problems may be forgotten later in training. The easy data resampling method addresses this by:
> > 
> > * Selectively sampling from the historical perfectly-solved list
> > * Reintroducing these problems to training
> > 
> > This effectively prevents catastrophic forgetting of previously learned solutions.
> 
> Thanks for the answer! Is there also a reason why the math filter is 90% while code is 100%?

The different thresholds actually come from the data distribution:

For math, we set the 90% filter because:

- We already had a large pool of math RL data (~150k additional samples above 90% accuracy). Including all these would skew the training distribution toward easier problems. So we kept the math threshold stricter to maintain balance. 

For code, we kept it at 100% because:

- Perfectly solved samples were relatively rare in the code RL data. Compared to the original math RL dataset, the code RL data was relatively small - containing only about 30K samples. Therefore, we didn't need to impose additional artificial limits.
Basically, the thresholds were tuned to keep both domains well-represented in training. Hope this explains the reasoning!

**Comment by faresobeid** [comment]
Oh ok that makes sense, thx!
